File: frontend/src/app/service/horse.service.ts
import {HttpClient, HttpParams} from '@angular/common/http';
import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {environment} from 'src/environments/environment';
import {Horse, HorseSearch} from '../dto/horse';

const baseUri = environment.backendUrl + '/horses';

@Injectable({
  providedIn: 'root'
})
export class HorseService {

  constructor(
    private http: HttpClient,
  ) {
  }

  /**
   * Get all horses stored in the system which match with the specified data
   *
   * @return observable list of found horses.
   */
  searchHorses(horseToSearch: HorseSearch): Observable<Horse[]> {
    let params = new HttpParams();

    // delete data if HorseSearch-data is empty or null
    // only add to params if not undefined
    if (horseToSearch.name?.length === 0 || horseToSearch.name == null) {
      delete horseToSearch.name;
    } else if (horseToSearch.name !== undefined) {
      params = params.append('name', horseToSearch.name);
    }

    if (horseToSearch.description?.length === 0 || horseToSearch.description == null) {
      delete horseToSearch.description;
    } else if (horseToSearch.description !== undefined) {
      params = params.append('description', horseToSearch.description);
    }

    if (horseToSearch.bornBefore?.length === 0 || horseToSearch.bornBefore == null) {
      delete horseToSearch.bornBefore;
    } else if (horseToSearch.bornBefore !== undefined) {
      params = params.append('bornBefore', horseToSearch.bornBefore);
    }

    if (horseToSearch.sex?.length === 0 || horseToSearch.sex == null) {
      delete horseToSearch.sex;
    } else if (horseToSearch.sex !== undefined) {
      params = params.append('sex', horseToSearch.sex);
    }

    if (horseToSearch.ownerName?.length === 0 || horseToSearch.ownerName == null) {
      delete horseToSearch.ownerName;
    } else if (horseToSearch.ownerName !== undefined) {
      params = params.append('ownerName', horseToSearch.ownerName);
    }

    return this.http.get<Horse[]>(baseUri, {params});
  }

  /**
   * Get a horse by its specified id
   *
   * @param id the id of the wanted horse
   * @return an Observable for the horse with specified id
   */
  get(id: number): Observable<Horse> {
    return this.http.get<Horse>(`${baseUri}/${id}`);
  }

  /**
   * Update an existing horse in the system
   *
   * @param horse the data for the horse that should be updated
   * @return an Observable for the updated horse
   */
  update(horse: Horse): Observable<Horse> {
    return this.http.put<Horse>(
      `${baseUri}/${horse.id}`,
      horse
    );
  }

  /**
   * Create a new horse in the system.
   *
   * @param horse the data for the horse that should be created
   * @return an Observable for the created horse
   */
  create(horse: Horse): Observable<Horse> {
    return this.http.post<Horse>(
      baseUri,
      horse
    );
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${baseUri}/${id}`);
  }

  public searchByName(name: string, limitTo: number): Observable<Horse[]> {
    const params = new HttpParams()
      .set('name', name)
      .set('maxAmount', limitTo);
    return this.http.get<Horse[]>(baseUri, {params});
  }

}

File: backend/src/main/java/at/ac/tuwien/sepm/assignment/individual/service/HorseService.java
package at.ac.tuwien.sepm.assignment.individual.service;

import at.ac.tuwien.sepm.assignment.individual.dto.HorseCreateDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseDetailDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseListDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseSearchDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseTreeDto;
import at.ac.tuwien.sepm.assignment.individual.exception.ConflictException;
import at.ac.tuwien.sepm.assignment.individual.exception.NotFoundException;
import at.ac.tuwien.sepm.assignment.individual.exception.ValidationException;

import java.util.stream.Stream;

/**
 * Service for working with horses.
 */
public interface HorseService {

  /**
   * Lists all horses stored in the system.
   *
   * @return list of all stored horses
   */
  Stream<HorseListDto> allHorses();

  /**
   * Lists all horses which match with the specified HorseSearchDto stored in the system.
   *
   * @return a list of all matching stored horses
   */
  Stream<HorseListDto> getSpecifiedHorses(HorseSearchDto horseSearchDto);


  /**
   * Updates the horse with the ID given in {@code horse}
   * with the data given in {@code horse}
   * in the persistent data store.
   *
   * @param horse the horse to update
   * @return the updated horse
   * @throws NotFoundException   if the horse with given ID does not exist in the persistent data store
   * @throws ValidationException if the update data given for the horse is in itself incorrect (description too long, no name, …)
   * @throws ConflictException   if the update data given for the horse is in conflict with the data currently in the system (owner does not exist, …)
   */
  HorseDetailDto update(HorseDetailDto horse) throws NotFoundException, ValidationException, ConflictException;


  /**
   * Get the horse with given ID, with more detail information.
   * This includes the owner of the horse, and its parents.
   * The parents of the parents are not included.
   *
   * @param id the ID of the horse to get
   * @return the horse with ID {@code id}
   * @throws NotFoundException if the horse with the given ID does not exist in the persistent data store
   */
  HorseDetailDto getById(long id) throws NotFoundException;

  /**
   * Create a new horse in the persistent data store.
   *
   * @param newHorse the data for the new horse
   * @return the horse, that was just newly created in the persistent data store
   * @throws NotFoundException   if the horse with given ID does not exist in the persistent data store
   * @throws ValidationException if the data given for the horse creation is in itself incorrect (description too long, no name, …)
   * @throws ConflictException   if the data given for the horse creation is in conflict with the data currently in the system (owner does not exist, …)
   */
  HorseDetailDto create(HorseCreateDto newHorse) throws NotFoundException, ValidationException, ConflictException;

  /**
   * Delete an existing horse from the persistent data store
   *
   * @param id the id of the horse which gets deleted
   * @throws NotFoundException if the horse with the given ID does not exist in the persistent data store
   */
  void delete(Long id) throws NotFoundException;

  /**
   * Get all ancestors of specified horse stored in the persistent data store until the specified generation.
   *
   * @param id             the ID of the horse to get the ancestors of
   * @param maxGenerations the maximum amount of generations which should get searched
   * @return the {@link HorseTreeDto} of the horse and its ancestors
   * @throws NotFoundException   if the horse with given ID does not exist in the persistent data store
   * @throws ValidationException if the data given for the horse-tree is incorrect (no id, no maxGenerations, invalid maxGenerations)
   */
  HorseTreeDto getAncestorHorses(long id, long maxGenerations) throws NotFoundException, ValidationException;
}
